drivers/sensor/bme280/bme280.c

```diff
@@ -306,95 +370,187 @@ static int bme280_chip_init(struct device *dev)
+	rc = bme280_chip_init(dev);
+	if (rc < 0) {
+		LOG_DBG("chip init failed: %d", rc);
 		return -EINVAL;
 	}
 
+	LOG_DBG("device %s initialized", dev->config->name);
 	return 0;
 }
 
-static struct bme280_data bme280_data;
 
-DEVICE_AND_API_INIT(bme280, DT_INST_0_BOSCH_BME280_LABEL, bme280_init, &bme280_data,
-		    NULL, POST_KERNEL, CONFIG_SENSOR_INIT_PRIORITY,
-		    &bme280_api_funcs);
+/* bme280_is_on_spi() can't be used at preprocessor time. */
+#define BME280_ON_SPI(inst) DT_INST_ON_BUS(inst, spi)
+
+/* This is only usable if BME280_ON_SPI(inst) */
+#define BME280_HAS_CS(inst) DT_INST_SPI_DEV_HAS_CS(inst)
+
+#define BME280_DEFINE(inst)						\
+	COND_CODE_1(BME280_ON_SPI(inst),				\
+		    (BME280_DEFINE_SPI(inst)),				\
+		    (BME280_DEFINE_I2C(inst)))
+
+#define BME280_DEFINE_SPI(inst)						\
+	static struct bme280_data bme280_data_##inst =			\
+		BME280_DATA_SPI(inst);					\
+	static const struct bme280_config bme280_config_##inst =	\
+		BME280_CONFIG_SPI(inst);				\
+	BME280_DEVICE_INIT(inst)
+
+#define BME280_DEFINE_I2C(inst)						\
+	static struct bme280_data bme280_data_##inst;			\
+	static const struct bme280_config bme280_config_##inst =	\
+		BME280_CONFIG_I2C(inst);				\
+	BME280_DEVICE_INIT(inst)
+
+#define BME280_DEVICE_INIT(inst)					\
+	DEVICE_AND_API_INIT(bme280_##inst,				\
+			    DT_INST_LABEL(inst),			\
+			    bme280_init,				\
+			    &bme280_data_##inst,			\
+			    &bme280_config_##inst,			\
+			    POST_KERNEL,				\
+			    CONFIG_SENSOR_INIT_PRIORITY,		\
+			    &bme280_api_funcs)
+
+#define BME280_DATA_SPI(inst)						\
+	COND_CODE_1(BME280_HAS_CS(inst),				\
+		    (BME280_DATA_SPI_CS(inst)),				\
+		    ({}))
+
+#define BME280_DATA_SPI_CS(inst)					\
+	{ .spi_cs = { .gpio_pin = DT_INST_SPI_DEV_CS_GPIO_PIN(inst), }, }
+
+#define BME280_CONFIG_SPI(inst)						\
+	{								\
+		.bus_label = DT_INST_BUS_LABEL(inst),			\
+		.reg_io = &bme280_reg_io_spi,				\
+		.bus_config = { .spi_cfg = BME280_SPI_CFG(inst)	}	\
+	}
+
+#define BME280_SPI_CFG(inst)						\
+	(&(struct bme280_spi_cfg) {					\
+		.spi_cfg = {						\
+			.frequency =					\
+				DT_INST_PROP(inst, spi_max_frequency),	\
+			.operation = (SPI_WORD_SET(8) |			\
+				      SPI_TRANSFER_MSB |		\
+				      SPI_MODE_CPOL |			\
+				      SPI_MODE_CPHA),			\
+			.slave = DT_INST_REG_ADDR(inst),		\
+			.cs = BME280_SPI_CS_PTR(inst),			\
+		},							\
+		.cs_gpios_label = BME280_SPI_CS_LABEL(inst),		\
+	})
+
+#define BME280_SPI_CS_PTR(inst)						\
+	COND_CODE_1(BME280_HAS_CS(inst),				\
+		    (&(bme280_data_##inst.spi_cs)),			\
+		    (NULL))
+
+#define BME280_SPI_CS_LABEL(inst)					\
+	COND_CODE_1(BME280_HAS_CS(inst),				\
+		    (DT_INST_SPI_DEV_CS_GPIO_LABEL(inst)), (NULL))
+
+#define BME280_CONFIG_I2C(inst)						\
+	{								\
+		.bus_label = DT_INST_BUS_LABEL(inst),			\
+		.reg_io = &bme280_reg_io_i2c,				\
+		.bus_config =  { .i2c_addr = DT_INST_REG_ADDR(inst), }	\
+	}
+
+/* TODO replace with a foreach over all enabled devices when available */
```

Comment:
@mbolivar-nordic when will the `foreach` macro be available?

Reply:
There is a PR. When it's merged it'll be ready :).

Reply:
seen #23416 thanks!